Initial implementation of the nested::Collapse for CUDA policies.

ajkunen opened the pull request:
This enables nested::forall to use the cuda_thread_*_exec and related policies from forallN inside of a CudaCollapse clause.

I've got this running with one unit test off the bat... I've got to try and get more testing established, but I wanted to get some feedback before I went gonzo.

trws commented on include/RAJA/internal/LegacyCompatibility.hpp:
```diff
@@ -180,6 +180,22 @@ RAJA_HOST_DEVICE RAJA_INLINE constexpr Result max(Args... args)
 }
 
 
+struct miner {
```
Any particular reason not to use RAJA::operators::minimum?

rhornung67 replied:
'miner'? Like a guy working underground with a pick and wearing a headlamp?  Can we find a better name for this; e.g., get_min, min_picker?

trws commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+using CudaCollapseAsync = Collapse<cuda_collapse_exec<true>, FOR ...>;
+
+
+// TODO, check that FT... are cuda policies
```
It may be worth just doing this, now that the check is actually pretty easy.  See [here](https://github.com/LLNL/RAJA/blob/develop/include/RAJA/pattern/forallN.hpp#L61).

trws commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+
+      // Assign the For index value to the correct argument
+      int loop_value = policy();
+      data.template assign_index<ft::index_val>( loop_value );
```
Ugh... that really needs a better interface doesn't it? I'll think about how to make this a little less cumbersome.

trws commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+
+
+    // Only launch a kernel if we have at least one thing to do
+    if (numBlocks(dims) > 0 && numThreads(dims) > 0) {
```
Any way we can make this a static assert for the case where there just aren't enough dimensions, in addition to the dynamic number of blocks/threads?

rhornung67 commented on include/RAJA/policy/cuda/forallN.hpp:
```diff
@@ -68,18 +68,6 @@ struct ForallN_BindFirstArg_Device {
 };
 
 
-RAJA_INLINE
```
Is this needed for some forallN use cases?

Never mind. I think you just moved it to someplace else.

rhornung67 commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+
+/*!
+ * \brief Launcher that uses execution policies to map blockIdx and threadIdx to
+ * map
```
remove line break here.

rhornung67 commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+
+  // Invoke the wrapped body.
+  // The wrapper will take care of computing indices, and deciding if the
+  // given block+thread is in-bounds, and invoking the users loop body
```
What happens if it's out of bounds? If it's a problem, it's worth notng here.

ajkunen replied:
There is no problem, it's just how CUDA is.  You always need to gaurd against this, unless your iteration space is exactly numBlocks*numThreads

rhornung67 commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+
+
+// TODO, check that FT... are cuda policies
+template <bool Async, typename ... FT>
```
What does 'FT' refer to? It's kinda cryptic.

rhornung67 commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+
+
+    /*
+     * Evaluates the loop index for the idx'th loop in the Collapse
```
idx'th? Wow. Try saying that ten times fast :-)

rhornung67 commented on test/unit/nested.cpp:
```diff
@@ -158,20 +158,20 @@ TEST(Nested, TileDynamic)
 
 
-#if 0 // NOT WORKING YET
+#if 1 // NOT WORKING YET
```
If it's not working, it should be turned off. Is the comment wrong?

ajkunen replied:
I just didn't want to pull the trigger

ajkunen commented:
@trws and @rhornung67  Thanks for the feedback, I've updated the branch a bit, and added the ability to run sequential loops within a cuda kernel.

I still need to get some more examples/tests together

rhornung67 commented:
@ajkunen when checks pass, I'll look it over and re-approve. I want to try some of this stuff out.

ajkunen commented:
@rhornung67 looks like the tests aren't running properly.. Did something with travis break?

rhornung67 commented:
@ajkunen I've seen things like this before. Not sure what the issue is. Try logging in to travis and manually restarting the build.

trws commented:
Restarted.

rhornung67 commented:
@trws I deleted the cache for @ajkunen 's PR and he restarted the build. Thanks.

rhornung67 commented on include/RAJA/internal/LegacyCompatibility.hpp:
```diff
@@ -145,40 +145,25 @@ RAJA_HOST_DEVICE RAJA_INLINE constexpr auto foldl(Op&& operation,
 RAJA_HOST_DEVICE RAJA_INLINE constexpr Result max(Args... args)
 {
-  return foldl(maxer(), args...);
+  return foldl(RAJA::operators::maximum<Result>(), args...);
```
Much better. 👍

rhornung67 commented on include/RAJA/policy/cuda/forall.hpp:
```diff
@@ -193,6 +193,32 @@ RAJA_INLINE void forall_impl(cuda_exec<BlockSize, Async>,
+                        Iterable&& iter,
+                        LoopBody&& loop_body)
+{
+  // TODO: we need a portable std::begin, std::end, and std::distance
```
Is this comment still relevant?

ajkunen replied:
Not sure.  I had to resort to using seg.begin() and seg.end() instead of std::begin(seg), std::end(seg).

I think in this context, it may not matter.

rhornung67 commented on include/RAJA/policy/cuda/nested.hpp:
```diff
@@ -113,30 +113,237 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
+  };
   template <typename WrappedBody>
-  void operator()(Collapse<cuda_collapse_exec, FT0, FT1> const &, WrappedBody const &wrap)
+  void RAJA_DEVICE operator()(ForType const &fp, WrappedBody const &wrap)
```
"ForType" Much better. 👍

ajkunen commented:
@rhornung67 I will do another PR to add more unit tests
